rpc/connection_log.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package rpc
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"runtime"
+	"strings"
+)
+
+const msgKey string = "msg"
+
+type logField struct {
+	key   string
+	value interface{}
+}
+
+// Format implements the fmt.Formatter interface, to make the structured
+// logField compatible with format-based non-structured loggers.
+func (f logField) Format(s fmt.State, verb rune) {
+	fmt.Fprintf(s, "%"+string(verb), f.value)
+}
+
+// connectionLog defines an interface used by connection.go for logging. An
+// implementation that does structural logging may ignore `format` completely
+// if `msgKey` is provided in logField.
+type connectionLog interface {
+	Warning(format string, fields ...logField)
+	Debug(format string, fields ...logField)
+	Info(format string, fields ...logField)
+}
+
+type connectionLogUnstructured struct {
+	LogOutput
+	logPrefix string
+}
+
+func newConnectionLogUnstructured(
+	logOutput LogOutput, prefix string) *connectionLogUnstructured {
+	randBytes := make([]byte, 4)
+	rand.Read(randBytes)
+	return &connectionLogUnstructured{
+		LogOutput: logOutput,
+		logPrefix: strings.Join(
+			[]string{prefix, hex.EncodeToString(randBytes)}, " "),
+	}
+}
+
+func formatLogFields(f string, lf ...logField) string {
+	fields := make([]interface{}, 0, len(lf))
+	for _, lf := range lf {
+		fields = append(fields, lf)
+	}
+	return fmt.Sprintf(f, fields...)
+}
+
+func (l *connectionLogUnstructured) Warning(
+	format string, fields ...logField) {
+	l.LogOutput.Warning("(%s) %s", l.logPrefix,
```

Comment:
Usually when we define a new `Debug`/`Warning`/`Error` method that wraps an existing one, we have to call `CloneWithAddedDepth` somewhere inside it, or else the log output will refer to this line instead of the line where logging is actually happening. Should we be doing that here?

Reply:
Oh good point thanks for bringing it up!